node/src/components/block_proposer.rs

```diff
@@ -197,251 +281,118 @@ impl BlockProposer {
-        }
-    }
-
-    /// Gets the chainspec from storage in order to call `propose_deploys()`.
-    fn get_chainspec_from_storage<REv: ReactorEventT>(
-        &mut self,
-        effect_builder: EffectBuilder<REv>,
-        chainspec_version: Version,
-        request: ListForInclusionRequest,
-    ) -> Effects<Event>
-    where
-        REv: From<StorageRequest> + Send,
-    {
-        effect_builder
-            .get_chainspec(chainspec_version.clone())
-            .event(move |maybe_chainspec| Event::GetChainspecResult {
-                maybe_deploy_config: Box::new(maybe_chainspec.map(|c| c.genesis.deploy_config)),
-                chainspec_version,
-                request,
-            })
-    }
-
-    /// Returns a list of candidates for inclusion into a block.
-    fn propose_deploys(
-        &mut self,
-        deploy_config: DeployConfig,
-        block_timestamp: Timestamp,
-        past_deploys: HashSet<DeployHash>,
-    ) -> HashSet<DeployHash> {
-        // deploys_to_return = all deploys in pending that aren't in finalized blocks or
-        // proposed blocks from the set `past_blocks`
-        self.state
-            .pending
-            .iter()
-            .filter(|&(hash, deploy)| {
-                self.is_deploy_valid(deploy, block_timestamp, &deploy_config, &past_deploys)
-                    && !past_deploys.contains(hash)
-                    && !self.state.finalized_deploys.contains_key(hash)
-            })
-            .map(|(hash, _deploy)| *hash)
-            .take(deploy_config.block_max_deploy_count as usize)
-            .collect::<HashSet<_>>()
-        // TODO: check gas and block size limits
-    }
-
-    /// Checks if a deploy is valid (for inclusion into the next block).
-    fn is_deploy_valid(
-        &self,
-        deploy: &DeployHeader,
-        block_timestamp: Timestamp,
-        deploy_config: &DeployConfig,
-        past_deploys: &HashSet<DeployHash>,
-    ) -> bool {
-        let all_deps_resolved = || {
-            deploy.dependencies().iter().all(|dep| {
-                past_deploys.contains(dep) || self.state.finalized_deploys.contains_key(dep)
-            })
-        };
-        let ttl_valid = deploy.ttl() <= deploy_config.max_ttl;
-        let timestamp_valid = deploy.timestamp() <= block_timestamp;
-        let deploy_valid = deploy.timestamp() + deploy.ttl() >= block_timestamp;
-        let num_deps_valid = deploy.dependencies().len() <= deploy_config.max_dependencies as usize;
-        ttl_valid && timestamp_valid && deploy_valid && num_deps_valid && all_deps_resolved()
-    }
-
     /// Notifies the block proposer that a block has been finalized.
     fn finalized_deploys<I>(&mut self, deploys: I)
     where
         I: IntoIterator<Item = DeployHash>,
     {
+        // TODO: This will ignore deploys that weren't in `pending`. They might be added
+        // later, and then would be proposed as duplicates.
```

Comment:
Good point! Is there a ticket for it?

Reply:
Not yet, as far as I know. @fizyk20 want to add one?